### PR TITLE
ooni/templates http: include all request/responses

### LINKED ISSUE #1143: web_connectivity: store all request/responses
Discussed with @hellais

### PR DESCRIPTION
**Pull request by bassosimone**
- make sure we include all requests and responses in OONI templates
  so that we see redirections in the results

- reduce the number of cases where issueing an HTTP request has
  as result an error coupled with a `nullptr` response

- fix regress to be compatible with both libevent 2.0 and 2.1

Closes #1143.

### CODE REVIEW

File: src/libmeasurement_kit/ooni/templates.cpp [hunk 1]
```diff
@@ -93,38 +93,51 @@ void http_request(Var<Entry> entry, Settings settings, http::Headers headers,
...
+                return rr;
+            };
 
-            if (!error) {
-                for (auto pair: response->headers) {
-                    rr["response"]["headers"][pair.first] =
-                        represent_string(redact(pair.second));
+            if (!!response) {
+                for (Var<http::Response> x = response; !!x; x = x->previous) {
+                    (*entry)["requests"].push_back(dump(x));
```
Comment (hellais) [inline src/libmeasurement_kit/ooni/templates.cpp]: From what I can tell each session pair (request, response) will always have the same request sent. So for example if I have the following chain of redirects:
```
-> GET http://example.com/one
<- Location: http://example.com/two
-> GET http://example.com/two
<- Location: http://other.example.com/three
-> GET http://other.example.com/three
<- Body: other.example.com/three
```

What I will read in the report will actually be:

```json
{
   "requests": [
      {
         "request": {"url": "http://example.com/one"},
         "response": {"body": "other.example.com/three"},
     },
     {
       "request": {"url": "http://example.com/one"},
       "response": {"location": "http://example.com/three"},
     },
     {
       "request": {"url": "http://example.com/one"},
       "response": {"location": "http://example.com/two"},
     },
  ]
}
```
Note: am I getting the ordering correct? (This is the ordering we use in ooniprobe, item `[0]` is the last response)

What I would have expected instead would have been:

```json
{
   "requests": [
      {
         "request": {"url": "http://example.com/three"},
         "response": {"body": "other.example.com/three"},
     },
     {
       "request": {"url": "http://example.com/two"},
       "response": {"location": "http://example.com/three"},
     },
     {
       "request": {"url": "http://example.com/one"},
       "response": {"location": "http://example.com/two"},
     },
  ]
}
```

Comment (bassosimone) [inline src/libmeasurement_kit/ooni/templates.cpp]: I've just run a web connectivity with input `https://httpbin.org/redirect/3` and this is an extract from the result NJSON file:

```JSON
{
    "requests": [
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/get"
            },
            "response": {
                "body": "{\n  \"args\": {}, \n  \"headers\": {\n    \"Accept\": \"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\", \n    \"Accept-Language\": \"en-US;q=0.8,en;q=0.5\", \n    \"Host\": \"httpbin.org\", \n    \"User-Agent\": \"Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36\"\n  }, \n  \"origin\": \"[REDACTED]\", \n  \"url\": \"https://httpbin.org/get\"\n}\n",
                "code": 200,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "385",
                    "Content-Type": "application/json",
                    "Date": "Tue, 28 Feb 2017 16:02:35 GMT",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 200 OK"
            }
        },
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/relative-redirect/1"
            },
            "response": {
                "body": "",
                "code": 302,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "0",
                    "Content-Type": "text/html; charset=utf-8",
                    "Date": "Tue, 28 Feb 2017 16:02:35 GMT",
                    "Location": "/get",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 302 FOUND"
            }
        },
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/relative-redirect/2"
            },
            "response": {
                "body": "",
                "code": 302,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "0",
                    "Content-Type": "text/html; charset=utf-8",
                    "Date": "Tue, 28 Feb 2017 16:02:34 GMT",
                    "Location": "/relative-redirect/1",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 302 FOUND"
            }
        },
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/redirect/3"
            },
            "response": {
                "body": "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\">\n<title>Redirecting...</title>\n<h1>Redirecting...</h1>\n<p>You should be redirected automatically to target URL: <a href=\"/relative-redirect/2\">/relative-redirect/2</a>.  If not click the link.",
                "code": 302,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "247",
                    "Content-Type": "text/html; charset=utf-8",
                    "Date": "Tue, 28 Feb 2017 16:02:34 GMT",
                    "Location": "/relative-redirect/2",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 302 FOUND"
            }
        }
    ]
}
```

Comment (hellais) [inline src/libmeasurement_kit/ooni/templates.cpp]: > Isn't this exactly what you would like to see?

Yes you are right that is exactly what I would like to see. My bad for not taking the time to actually test it and just relying on what I could infer from the code.

**Comment by hellais** [comment]
Steps:

* [x] Review diff
* [ ] Local smoke testing
* [ ] Approve

File: src/libmeasurement_kit/ooni/templates.cpp [hunk 2]
```diff
@@ -93,38 +93,51 @@ void http_request(Var<Entry> entry, Settings settings, http::Headers headers,
...
-            for (auto pair: headers) {
-                rr["request"]["headers"][pair.first] =
-                    represent_string(redact(pair.second));
-            }
-            rr["request"]["body"] = represent_string(redact(body));
-            rr["request"]["url"] = settings.at("http/url").c_str();
-            rr["request"]["method"] = settings.at("http/method").c_str();
 
-            // XXX we should probably update OONI data format to remove this.
-            rr["method"] = settings.at("http/method").c_str();
```
Comment (bassosimone) [inline src/libmeasurement_kit/ooni/templates.cpp]: I don't remember history either. Anyway, I checked what ooni-probe produces, noticed it does not include that line, and then decided it was safe to ditch it and the related comment.

File: src/libmeasurement_kit/ooni/templates.cpp (hunk 1 above)
Comment (bassosimone) [inline src/libmeasurement_kit/ooni/templates.cpp]: The catch is that, before this diff, the `request` in the report was filled using `settings`. After this diff, instead, we use the `request` object contained by a `response` to fill the report. This is why, incidentally, this diff also intervenes in the `http` module to make sure a valid, albeit incomplete, `response` is returned in most cases in the error path (except when of course it does not make sense to do so).

**Comment by hellais** [comment]
I did some local testing. The output looks consistent and it's working as expected.

I had to re-run the tests a couple of times to make them work, but I suspect that is due to temporary network failures.

**Comment by bassosimone** [comment]
> I had to re-run the tests a couple of times to make them work, but I suspect that is due to temporary network failures.

What tests were failing?

**Comment by bassosimone** [comment]
Speaking of myself in third person, @bassosimone will now backport this diff to the stable branch such that we can tag v0.4.1. Thank you for the review!

**Pull request merged**